src/algorithms/split_and_merge_algorithm.cc

```diff
@@ -0,0 +1,408 @@
+    compute_C_launch(i,j);
+
+    // Restricted GS steps
+    for(unsigned int t=0; t<T; ++t){
+      restricted_GS(i,j);
+    }
+
+    split_or_merge(i,j);
+  }
+
+  // Full GS updates
+  for(unsigned int k=0; k<M; ++k){
+    full_GS();
+  }
+  
+}
+
+void SplitAndMergeAlgorithm::sample_unique_values(){
+  for (auto &un : unique_values) {
+    un->sample_full_cond(!update_hierarchy_params());
+  }
+}
+
+Eigen::VectorXd SplitAndMergeAlgorithm::lpdf_marginal_component(
+    std::shared_ptr<AbstractHierarchy> hier, const Eigen::MatrixXd &grid,
+    const Eigen::RowVectorXd &covariate) const {
+  return hier->prior_pred_lpdf_grid(grid, covariate);
+}
+
+void SplitAndMergeAlgorithm::compute_S(const unsigned int i, const unsigned int j) {
+    unsigned int lengthS=0;
+    for(int k=0; k<allocations.size(); k++){
+        if(allocations[k]==allocations[i]||allocations[k]==allocations[j])
+            lengthS++;
+    }
+    lengthS=lengthS-2;
+    S={}; //Necessary, since it needs to be zero at each iter, not only at the beginning
+    S.resize(lengthS,0);
+    unsigned int index=0;
+    for (int k = 0; k < allocations.size(); ++k) {
+        if ((allocations[k]==allocations[i]||allocations[k]==allocations[j])&& (k!=i) && (k!=j)) {
+            S[index]=k;
+            index++;
+        }
+    }
+}
+
+void SplitAndMergeAlgorithm::compute_C_launch(const unsigned int i,
+  const unsigned int j){
+  allocations_cl.clear();
+  allocations_cl.resize(S.size());
+
+  cl.clear();
+  cl.push_back(unique_values[0]->clone());
+  cl.push_back(unique_values[0]->clone());
+  
+  auto &rng = bayesmix::Rng::Instance().get(); 
+  Eigen::VectorXd probas(2);
+  probas(0)=0.5;
+  probas(1)=0.5;
+  for (size_t k = 0; k < S.size(); ++k) {
+    if (bayesmix::categorical_rng(probas, rng, 0)){
+      allocations_cl[k]=1;
+      cl[1]->add_datum(S[k], data.row(S[k]), false);
+    }else{
+      allocations_cl[k]=0;
+      cl[0]->add_datum(S[k], data.row(S[k]), false);
+    }
+  }
+  /* We update the posterior parameters only at the last insertion to ease
+   * computations.
+   */
+  cl[0]->add_datum(i, data.row(i), true);
+  cl[1]->add_datum(j, data.row(j), true);
+}
+
+void SplitAndMergeAlgorithm::split_or_merge(const unsigned int i, const unsigned int j){
+  if(allocations[i]==allocations[j]){  
+    std::vector<unsigned int> clSplit (allocations.size()); 
+    std::shared_ptr<AbstractHierarchy> data_i(unique_values[0]->clone());
+    std::shared_ptr<AbstractHierarchy> data_j(unique_values[0]->clone());  
+    std::shared_ptr<AbstractHierarchy> data_J(unique_values[0]->clone());
+    double p_i=0; 
+    double p_j=0;
+    double p_J=0;
+    unsigned int LabI = unique_values.size();
+    
+    clSplit[i]=LabI;
+    clSplit[j]=allocations[j];
+    double q=std::exp(restricted_GS(i,j,true));
+    unsigned int z=0;
+    std::set<int> set_i=cl[0]->get_data_idx();
+    std::set<int> set_j=cl[1]->get_data_idx();
+    std::set<int>::const_iterator iter_i=set_i.cbegin();
```

Comment:
I don't think you actually need an iterator, since you're only using *iter in the rest of the code. It's cleaner to let these `int i_idx; j_idx`